File: Leetcode_Code/17. Letter Combinations of a Phone Number/17_dfs.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.length() == 0) return result;
        vector<string> dict = {" ", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        letterCombinations_dfs(digits, "", 0, dict);
        return result;
    }

private:
    vector<string> result;

    void letterCombinations_dfs(string &digits, string now, int index, vector<string> &dict) {
        if (index == digits.length()) {
            result.push_back(now);
            return;
        }
        string letter = dict[digits[index] - '0'];
        for (int i = 0; i < letter.length(); ++i)
            letterCombinations_dfs(digits, now + letter[i], index + 1, dict);
    }
};

File: Leetcode_Code/17. Letter Combinations of a Phone Number/17_dfs.py
class Solution(object):
    def __init__(self):
        self.result = []
        self.dict = [" ", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits) == 0:
            return self.result
        self.letterCombinations_dfs(digits, "", 0)
        return self.result

    def letterCombinations_dfs(self, digits, now, index):
        if index == len(digits):
            self.result.append(now)
            return
        letter = self.dict[int(digits[index])]
        for i in range(len(letter)):
            self.letterCombinations_dfs(digits, now + letter[i], index + 1)
